Export CSV

subdavis opened the pull request:
WIP, ignore this.

subdavis commented:
There's a problem here.

store's datasets are only updated on load, but the confidence info (from meta) is needed at download time. If a user loads, changes the confidence, and then tries to download, it won't have updated.

This highlights the need for a couple changes

- [ ] save dataset metadata json individually rather than in 1 big array.  The cost of rewriting the entire recents array into local storage will become too large, especially when the user is changing confidences.
- [ ] Add a new ID array (just like the original).  So in order to load meta about a dataset, 2 queries into local storage would be needed.

Rather than doing a simple fix and changing local storage later when we might have to migrate, I think it might be best to just do it now.  Feedback welcome, though.

BryonLewis commented:
> Rather than doing a simple fix and changing local storage later when we might have to migrate, I think it might be best to just do it now. Feedback welcome, though.

Are you avoiding reading the confidence filter value directly from the meta.json file using an endpoint/ or subset of reading the file for performance reasons (It's a onetime cost during export for each project and ensures you have the latest data)?  We write to it frequently enough when editing the confidenceFilter.  I'm asking because if we want `localStorage` to be the official reading place of metadata values this would impact attributes and how they are read from as well as edited to keep them in sync with the meta.json file.

subdavis commented:
> Are you avoiding reading the confidence filter value directly from the meta.json file using an endpoint/ or subset of reading the file for performance reasons

Fair point.  I could perform a full metadata load when the dialog is opened.    

> I'm asking because if we want localStorage to be the official reading place of metadata values this would impact attributes and how they are read from as well as edited to keep them in sync with the meta.json file.

I like having a cheap way to get dozens of metadata objects at once without reading them all from disk, mostly for populating the recents and training pages.  

So far, I'm treating local storage as a possibly invalid cache. I would never use it to do anything important like mutation or a viewer load, but it's handy to have for listing operations.  Maybe there's a way to keep only a subset of the immutable meta fields in this cache.

Either way, I like your idea, I won't make any cchanges to the dataset store in this PR.

BryonLewis reviewed (changes requested):
A couple of minor questions and comments.
Seems to export and work well including all the necessary attributes (shouldn't be surprised we are using it for pipelines and training anyways).

subdavis commented on client/platform/desktop/frontend/components/Export.vue:
```diff
@@ -0,0 +1,151 @@
+              hide-details
+            />
+            <div
+              v-if="data.meta && data.meta.confidenceFilters"
+              class="py-2"
+            >
+              <div>Current thresholds:</div>
+              <span
+                v-for="(val, key) in data.meta.confidenceFilters"
+                :key="key"
+                class="pt-2"
+              >
+                ({{ key }}, {{ val }})
```
I didn't like the idea of having more than one source of truth for that value, and importing it from `useTrackFilters` would be inconsistent since the default is established at runtime.

Should do something about this, but none of the options seemed appropriate, so I ignored it.

subdavis commented on client/platform/desktop/frontend/api.ts:
```diff
@@ -75,6 +75,20 @@ async function importMedia(path: string): Promise<JsonMeta> {
 }
 
+async function exportDataset(id: string, exclude: boolean): Promise<string> {
+  const location = await remote.dialog.showSaveDialog({
+    title: 'Export Dataset',
+    defaultPath: `result_${id}.csv`,
```
yes

subdavis commented on client/platform/desktop/frontend/components/Export.vue:
```diff
@@ -0,0 +1,151 @@
+        </v-card-title>
+
+        <v-card-text class="pb-0">
+          <v-alert
+            v-if="data.err"
+            type="error"
+          >
+            {{ data.err }}
+          </v-alert>
```
There's really nothing that can be done to make that much text fit in this window.   I didn't want to use the prompt service.  

I guess I'll open another dialog.  Export errors are important, so a whole dialog seems appropriate.

subdavis commented on client/platform/desktop/frontend/components/ViewerLoader.vue:
```diff
@@ -29,6 +32,13 @@ export default defineComponent({
 
 <template>
   <Viewer :id="id">
+    <template #extension-right>
+      <Export
+        v-if="datasets[id]"
+        :id="id"
+        :small="true"
+      />
+    </template>
```
![Screenshot from 2021-02-01 13-16-21](https://user-images.githubusercontent.com/4214172/106500335-b08c0780-648f-11eb-9864-ed38a7d38088.png)

Arbitrary choice.  On small displays, something will always suffer.  Can change it if you want.

The above looked nicer to me.

subdavis replied:
TBH I think we can get rid of "DIVE" on the left hand side for desktop.  It serves no real purpose.

BryonLewis replied:
![image](https://user-images.githubusercontent.com/61746913/106500524-f052ef00-648f-11eb-9ca0-850b85820761.png)
I just checked because I forgot, we do have the download button next to pipelines in the web version.  I truly don't feel strongly about it one way or another so whatever you decide is fine with me.

subdavis commented:
![Screenshot from 2021-02-01 13-51-33](https://user-images.githubusercontent.com/4214172/106505346-30b56b80-6496-11eb-97ee-0f27e54a85a3.png)

Moved button to main toolbar, changed to home default location.

BryonLewis reviewed (approved):
Looks good.
